split up pandas/tests/indexes/test_multi.py #18644

elmq0022 opened the pull request:
- [x] closes #18644
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

The file is split out, the number of tests remains 377 (2 skips on my machine; all other test cases pass), the code uses fixtures in all instances, new fixtures are in a local conftest.py file, and there are no dependencies on classes.  

I'm happy to receive and address any feedback.

elmq0022 committed:
split up pandas/tests/indexes/test_multi.py #18644

jreback commented on pandas/tests/indexes/multi/conftest.py:
```diff
@@ -0,0 +1,42 @@
+
+
+@pytest.fixture
+def _index():
```
you don't need to prefix with '_'

elmq0022 replied:
The non-fixture variable "index" is used in many of these tests.  Several of these also use the fixture "index" which I dubbed _index in an attempt to distinguish them.  

I suggest using something like m_index or index_fixture to provide some separation. 

Thoughts?

jreback commented on pandas/tests/indexes/multi/test_base.py:
```diff
@@ -0,0 +1,938 @@
+from pandas.core.indexes.datetimelike import DatetimeIndexOpsMixin
+
+
+def verify_pickle(indices):
```
don't create functions like this, inline it instead (unless there is a lot of usage about this, but indicate that)

jreback commented on pandas/tests/indexes/multi/test_base.py:
```diff
@@ -0,0 +1,938 @@
+    # GH11193, when an existing index is passed, and a new name is not
+    # specified, the new index should inherit the previous object name
+    expected = _index
+    if not isinstance(expected, MultiIndex):
```
why would this NOT be a MI?

jreback commented on pandas/tests/indexes/multi/test_base.py:
```diff
@@ -0,0 +1,938 @@
+# -*- coding: utf-8 -*-
+
```
as long as you are splitting, would not create a test_base, rather create functional named files, similar to how ``pandas/tests/indexes/datetimes``; since you already have some functional area files below, just distribute these

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,900 @@
+    assert not _index.is_numeric()
+
+
+def test_bounds(_index):
```
where did this come from?

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,900 @@
+    expected = mi.get_level_values(level)
+
+
+def test_multiindex_compare():
```
move compare/equal/identical to a new file

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,900 @@
+
+    idx = MultiIndex.from_arrays([['qux', 'baz', 'foo', 'bar'], np.arange(
+        4)])
+    result = idx.isin(values)
```
isin should be in indexing

jreback commented on pandas/tests/indexes/multi/test_reindex.py:
```diff
@@ -0,0 +1,88 @@
+                           level='first')
+
+
+def test_reindex_preserves_names_when_target_is_list_or_ndarray(_index):
```
call this test_indexing

elmq0022 commented:
@jreback thank you for the quick feedback.  I will make updates over the next day or two.

elmq0022 committed:
addressing comments to pull request #21514 for GH issue #18644

elmq0022 committed:
Changes fixture _index to fixture idx

elmq0022 committed:
linting after fixes

jreback commented on pandas/tests/indexes/multi/test_integrity.py:
```diff
@@ -0,0 +1,448 @@
+        assert result == i
+
+
+def test_dims():
```
next PR can come back and remove this

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,748 @@
+        first.tolist() - idx[-3:]
+
+
+def test_nlevels(idx):
```
should be in integrity

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,748 @@
+        # GH 21149
+        # Ensure comparison operations for MultiIndex with nlevels == 1
+        # behave consistently with those for MultiIndex with nlevels > 1
+
```
move to equivalence

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,748 @@
+    tm.assert_raises_regex(ValueError, "Duplicated level name:",
+                           mi.rename, names[:2], level=[0, 2])
+
+
```
can you split out duplicated & uniques tests to a separate file

jreback commented on pandas/tests/indexes/multi/test_sorting.py:
```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+import numpy as np
+import pandas.util.testing as tm
```
I think you have a couple of tests for unsorted in integrity, can you move here

jreback commented on pandas/tests/indexes/multi/conftest.py:
```diff
@@ -0,0 +1,42 @@
+    return index
+
+
+@pytest.fixture
```
I see that you are iterating over this in a number of tests. why is that?

elmq0022 replied:
Okay, done with that, but some tests are not relevant to a the MultiIndex class or would need to be refactored.  I marked those with a TODO.  Those that could refactor, I did.  :sweat_smile:

jreback replied:
yep that's ok. i expect after your refactor will need to go over this again and clean a bit

jreback reviewed (changes requested):
looking really good. a couple of comments / questions.

elmq0022 committed:
moved tests according to feedback

test_nlevels -> test_integrity.py
unsort tests -> test_sorting.py
duplicates and unique test -> test_unqi_dups.py

elmq0022 committed:
removing the named_index fixture

jreback commented on pandas/tests/indexes/multi/conftest.py:
```diff
@@ -0,0 +1,37 @@
+
+@pytest.fixture
+def idx():
+    major_axis = Index(['foo', 'bar', 'baz', 'qux'])
```
can you add a comment describing the fixture (always do this for fixtures)

jreback commented on pandas/tests/indexes/multi/conftest.py:
```diff
@@ -0,0 +1,37 @@
+    return ['first', 'second']
+
+
+@pytest.fixture
```
ok w/o using these a leading _ (e.g. holder) is ok

jreback commented on pandas/tests/indexes/multi/test_compat.py:
```diff
@@ -0,0 +1,129 @@
+    def f():
+        if idx:
+            pass
+
```
can you change to do this with ``with``

elmq0022 replied:
I'm not following here.  What is the need to use a context manager here?  None of the functions seems to have a valid __enter__ or __exit__ methods.

jreback replied:
yep exactly

jreback commented on pandas/tests/indexes/multi/test_compat.py:
```diff
@@ -0,0 +1,129 @@
+    def f():
+        if common:
+            pass
+
```
same here

elmq0022 replied:
same as above

jreback commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -0,0 +1,510 @@
+
+
+def test_reconstruct_sort():
+
```
these 2 should be in test_sorting (reconstrut_*)

jreback commented on pandas/tests/indexes/multi/test_contains.py:
```diff
@@ -0,0 +1,129 @@
+def test_hasnans_isnans(idx):
+    # GH 11343, added tests for hasnans / isnans
+    # TODO: remove or change test not valid for MultiIndex
+    if isinstance(idx, MultiIndex):
```
did you comment?

can you leave the code and just xfail this test

elmq0022 replied:
I removed the comments, refactored to run original code, and marked this test to xfail.

jreback commented on pandas/tests/indexes/multi/test_conversion.py:
```diff
@@ -0,0 +1,150 @@
+    assert result.names == index.names
+
+
+def test_legacy_pickle():
```
can you use the skip decorator here

elmq0022 replied:
Done

jreback commented on pandas/tests/indexes/multi/test_copy.py:
```diff
@@ -0,0 +1,124 @@
+
+
+def test_copy_and_deepcopy(indices):
+    from copy import copy, deepcopy
```
can you move imports to the top

elmq0022 replied:
Done

jreback commented on pandas/tests/indexes/multi/test_equivalence.py:
```diff
@@ -0,0 +1,253 @@
+def test_equals_op(idx):
+    # GH9947, GH10637
+    index_a = idx
+    if isinstance(index_a, PeriodIndex):
```
prob can remove this (the isinstance check)

elmq0022 replied:
Done

jreback commented on pandas/tests/indexes/multi/test_equivalence.py:
```diff
@@ -0,0 +1,253 @@
+    assert not idx.is_numeric()
+
+
+def test_nulls(idx):
```
do we have a test_missing? if so pls move there

jreback commented on pandas/tests/indexes/multi/test_get_set.py:
```diff
@@ -0,0 +1,649 @@
+
+
+@pytest.mark.parametrize('level', [0, 1])
+@pytest.mark.parametrize('null_val', [np.nan, pd.NaT, None])
```
we have a null fixture that does this (for the nans u r listing)

elmq0022 replied:
Using the fixture `nulls_fixture` increases the number of nulls (and tests) by four, from 377 to 381.

jreback commented on pandas/tests/indexes/multi/test_get_set.py:
```diff
@@ -0,0 +1,649 @@
+
+    arrays = [[], []]
+    index = pd.MultiIndex.from_arrays(arrays)
+    result = index.get_level_values(0)
```
so out of this pls move to test_indexing:

get_indexer, getitem and like. this test should be for get/set levels, labels, names, get_level_values and so on

elmq0022 replied:
I think this one is done, but please go over it one last time.

elmq0022 replied:
I also moved `test_get_unique_index` to test_uniq_dups.

jreback replied:
i indicated some more to move

jreback commented on pandas/tests/indexes/multi/test_indexing.py:
```diff
@@ -0,0 +1,221 @@
+    assert s.name == idx.name
+
+
+def test_to_series_with_arguments(idx):
```
move to_series to test_conversion

elmq0022 replied:
Done.

jreback commented on pandas/tests/indexes/multi/test_indexing.py:
```diff
@@ -0,0 +1,221 @@
+        result = idx.delete(len(idx))
+
+
+def test_fillna(idx):
```
can you make a test_missing, move fillna, dropna there

elmq0022 replied:
Done

jreback commented on pandas/tests/indexes/multi/test_integrity.py:
```diff
@@ -0,0 +1,403 @@
+                           indices, mode='clip')
+
+
+def test_isna_behavior(idx):
```
move to test_missing (anything na related is fair game for that)

elmq0022 replied:
Done.

jreback commented on pandas/tests/indexes/multi/test_integrity.py:
```diff
@@ -0,0 +1,403 @@
+    assert not ind.is_monotonic
+
+
+def test_partial_string_timestamp_multiindex():
```
partial string should be in test_indexing, or test_timeseries

elmq0022 replied:
I moved this to test_indexing.

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,491 @@
+@pytest.mark.parametrize('first_type,second_type', [
+    ('int64', 'int64'),
+    ('datetime64[D]', 'str')])
+def test_remove_unused_levels_large(first_type, second_type):
```
move this to test_sorting

elmq0022 replied:
Done.

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,491 @@
+                                    ['a', 'd', 'b', 'unused']])
+@pytest.mark.parametrize('level1', [['w', 'x', 'y', 'z'],
+                                    ['w', 'x', 'y', 'z', 'unused']])
+def test_remove_unused_nan(level0, level1):
```
move to test_sorting

elmq0022 replied:
I moved it, but why not in test_missing.py?  Is it because of the implementation of the underlying algorithm?

jreback commented on pandas/tests/indexes/multi/test_uniq_dups.py:
```diff
@@ -0,0 +1,236 @@
+         (u('x'), u('out'), u('z'), 7, u('y'), u('in'), u('z'), 119),
+         (u('x'), u('out'), u('z'), 9, u('y'), u('in'), u('z'), 135),
+         (u('x'), u('out'), u('z'), 13, u('y'), u('in'), u('z'), 145),
+         (u('x'), u('out'), u('z'), 14, u('y'), u('in'), u('z'), 158),
```
just rename module to: test_unique_and_duplicates

elmq0022 replied:
Done

jreback commented on pandas/tests/indexes/multi/test_set_ops.py:
```diff
@@ -0,0 +1,276 @@
+    assert idx[:0].empty
+
+
+def test_unique_na():
```
move test_unique_and_duplicates

jreback reviewed (changes requested):
a number of comments, but should be straightforward. ping on green.

elmq0022 committed:
split test_multi.py into separate files and address the comments from pull request

jreback commented on pandas/tests/indexes/multi/test_contains.py:
```diff
@@ -0,0 +1,128 @@
+
+
+@pytest.mark.xfail
+def test_hasnans_isnans(idx):
```
can you move to test_missing

jreback commented on pandas/tests/indexes/multi/test_contains.py:
```diff
@@ -0,0 +1,128 @@
+    if len(idx) == 0:
+        pass
+    elif isinstance(idx, DatetimeIndexOpsMixin):
+        values[1] = iNaT
```
can simplify this here

jreback commented on pandas/tests/indexes/multi/test_conversion.py:
```diff
@@ -0,0 +1,179 @@
+    assert result.names == index.names
+
+
+@pytest.mark.skip
```
needs to be ``@pytest.mark.skipif(PY3, reason=......)``

jreback commented on pandas/tests/indexes/multi/test_get_set.py:
```diff
@@ -0,0 +1,514 @@
+from pandas.compat import lrange, range
+
+
+def test_get_loc(idx):
```
get_loc* also goes to test_indexing

jreback commented on pandas/tests/indexes/multi/test_get_set.py:
```diff
@@ -0,0 +1,514 @@
+
+    pytest.raises(KeyError, idx.get_loc, (False, True))
+    pytest.raises(KeyError, idx.get_loc, (True, False))
+
```
thru here

jreback commented on pandas/tests/indexes/multi/test_indexing.py:
```diff
@@ -0,0 +1,347 @@
+    assert result == (0, len(index))
+
+
+def test_shift(idx):
```
move to test_analytics

elmq0022 replied:
From the "I indicated some more to move" above, my impression is to move all the test_get_level_* to test_indexing.py.  Is that correct?

jreback commented on pandas/tests/indexes/multi/test_indexing.py:
```diff
@@ -0,0 +1,347 @@
+    assert indexer.dtype == np.intp
+
+
+def test_partial_string_timestamp_multiindex():
```
can you move test_partial  to a new file: test_partial_indexing.py

jreback commented on pandas/tests/indexes/multi/test_integrity.py:
```diff
@@ -0,0 +1,310 @@
+
+
+def test_nan_stays_float():
+
```
move to test_missing

jreback commented on pandas/tests/indexes/multi/test_missing.py:
```diff
@@ -0,0 +1,121 @@
+                tm.assert_numpy_array_equal(index.notna(), ~result)
+
+
+@pytest.mark.parametrize('level', [0, 1])
```
these last 2 here move to test_indexing

elmq0022 commented:
So, someone removed the get_data_path from the pandas.utils.testing.  I forgot to run pytest after I rebased.  I can inline the function if need be.  

The bigger issue is the rebase broke my ability to run pytest even on master.  

```
Traceback (most recent call last):
  File "/home/zero/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/_pytest/config.py", line 334, in _getconftestmodules
    return self._path2confmods[path]
KeyError: local('/home/zero/pandas/pandas/tests/indexes/test_multi.py')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/zero/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/_pytest/config.py", line 334, in _getconftestmodules
    return self._path2confmods[path]
KeyError: local('/home/zero/pandas/pandas/tests/indexes')
During handling of the above exception, another exception occurred:
Traceback (most recent call last):  File "/home/zero/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/_pytest/config.py", line 365, in _importconftest
    return self._conftestpath2mod[conftestpath]
KeyError: local('/home/zero/pandas/pandas/conftest.py')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/zero/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/_pytest/config.py", line 371, in _importconftest
    mod = conftestpath.pyimport()
  File "/home/zero/anaconda3/envs/pandas-dev/lib/python3.6/site-packages/py/_path/local.py", line 668, in pyimport
    __import__(modname)
  File "/home/zero/pandas/pandas/__init__.py", line 42, in <module>
    from pandas.core.api import *
  File "/home/zero/pandas/pandas/core/api.py", line 10, in <module>
    from pandas.core.groupby.groupby import Grouper
  File "/home/zero/pandas/pandas/core/groupby/__init__.py", line 2, in <module>
    from pandas.core.groupby.groupby import (
  File "/home/zero/pandas/pandas/core/groupby/groupby.py", line 46, in <module>
    from pandas.core.index import (Index, MultiIndex,
  File "/home/zero/pandas/pandas/core/index.py", line 2, in <module>
    from pandas.core.indexes.api import *
  File "/home/zero/pandas/pandas/core/indexes/api.py", line 4, in <module>
    from pandas.core.indexes.base import (Index,
  File "/home/zero/pandas/pandas/core/indexes/base.py", line 7, in <module>
    from pandas._libs import (lib, index as libindex, tslib as libts,
  File "pandas/_libs/index.pyx", line 28, in init pandas._libs.index
    from pandas._libs.tslibs import period as periodlib
  File "pandas/_libs/tslibs/period.pyx", line 59, in init pandas._libs.tslibs.period
  File "/home/zero/pandas/pandas/tseries/offsets.py", line 2327, in <module>
    offset=BDay(), time_rule=None):
  File "/home/zero/pandas/pandas/tseries/offsets.py", line 484, in __init__
    BaseOffset.__init__(self, n, normalize)
TypeError: object.__init__() takes no parameters
ERROR: could not load /home/zero/pandas/pandas/conftest.py
```

WillAyd commented:
You need to rebuild the Cython extensions after pulling from master as there have been recent changes with those

jreback reviewed (changes requested):
last round! thanks for this. ping on green.

elmq0022 commented:
Will do thanks!

On Tue, Jun 26, 2018 at 12:52 PM, William Ayd <notifications@github.com>
wrote:

> You need to rebuild the Cython extensions after pulling from master as
> there have been recent changes with those
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/21514#issuecomment-400405575>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADSn4F5qlxelBvmPjD6W-vs43sPeqq8Iks5uAnTkgaJpZM4Uq6xt>
> .
>

elmq0022 commented:
I think the datapath fixture is a step backward from the get_data_path which was relative to the test directory.  I can't run pytest from the directly from the tests/indexes/mulit directory successfully anymore.  I now have to run it from the root directory if I want the test to pass.  This should probably be relative to the director of the test like before.

jreback commented:
> I think the datapath fixture is a step backward from the get_data_path which was relative to the test directory. I can't run pytest from the directly from the tests/indexes/mulit directory successfully anymore. I now have to run it from the root directory if I want the test to pass. This should probably be relative to the director of the test like before.

hmm. I almost *always* run at the root directory. but I suppose this could be addressed as well. can you open a separate issue for this?

cc @TomAugspurger

TomAugspurger commented:
How does pytest handle fixtures when you don't run from the root of the tests dir?

elmq0022 commented:
I don't think it is necessarily a pytest issue.  An absolute path would
just be better since it makes the working directory  irrelevant.  Something
like os.path.dir(__file__) to resolve this would be relatively straight
forward.

On Tue, Jun 26, 2018, 2:18 PM Tom Augspurger <notifications@github.com>
wrote:

> How does pytest handle fixtures when you don't run from the root of the
> tests dir?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/21514#issuecomment-400431297>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADSn4OBcTwxX3X8AMs7wiAADqPtZdpeGks5uAokIgaJpZM4Uq6xt>
> .
>

TomAugspurger commented:
We need a fixture so that these tests are skipped / failed (depending on
the config option) when the file isn't present.

What exception do you get?



On Tue, Jun 26, 2018 at 2:38 PM, LeakedMemory <notifications@github.com>
wrote:

> I don't think it is necessarily a pytest issue. An absolute path would
> just be better since it makes the working directory irrelevant. Something
> like os.path.dir(__file__) to resolve this would be relatively straight
> forward.
>
> On Tue, Jun 26, 2018, 2:18 PM Tom Augspurger <notifications@github.com>
> wrote:
>
> > How does pytest handle fixtures when you don't run from the root of the
> > tests dir?
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/pandas-dev/pandas/pull/21514#issuecomment-400431297
> >,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> ADSn4OBcTwxX3X8AMs7wiAADqPtZdpeGks5uAokIgaJpZM4Uq6xt>
> > .
>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/21514#issuecomment-400436956>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIgH1RRF_RYFFOQHTzMMM__rMXUEbks5uAo2qgaJpZM4Uq6xt>
> .
>

elmq0022 commented:
I will try to get to the last round of comments later this week.  I pushed some fixes prior to seeing the comments.

elmq0022 commented:
@TomAugspurger  I see where you're going with the fixture.  The approach definitely makes sense, but I just like that absolute file paths let me run code regardless of where I am on the command line.  I opened an issue, #21646, and submitted a pull request to address it.  Thanks.

jreback commented:
thanks @elmq0022 ping when you want me to look.

elmq0022 commented:
@jreback looks like everything is green.  I believe all the comments are addressed.  I also included the last updates to test_multi.py in test unique and duplicates.

elmq0022 commented:
@jreback it looks like there were a couple of tests added that I need to include and rebase.  Do you have any additional feedback.  I would like to get this merged as soon as possible so everything can stay in sync.  Thanks!

jreback commented:
yes I agree. pls rebase. ping as soon as green and can merge.

jreback commented:
great thanks! I will go thru and  possibly make some comments. but take them as for a followup.

jreback commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -0,0 +1,434 @@
+        tm.assert_index_equal(result, expected)
+
+
+def test_from_arrays_invalid_input():
```
parameterize

jreback commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -0,0 +1,434 @@
+        pytest.raises(TypeError, MultiIndex.from_arrays, arrays=i)
+
+
+def test_from_arrays_different_lengths():
```
parameterize

jreback commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -0,0 +1,434 @@
+                              labels=[[], [], []], names=names)
+        tm.assert_index_equal(result, expected)
+
+
```
parametrize

elmq0022 replied:
Broke into several parameterized tests as the length of the names and the levels were different in each.  Trying to do it all in one level seemed clunky to me.

jreback commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -0,0 +1,434 @@
+
+
+def test_from_product_index_series_categorical():
+    # GH13743
```
parametrize

jreback commented on pandas/tests/indexes/multi/test_constructor.py:
```diff
@@ -0,0 +1,434 @@
+    index = idx
+    index.names = ['foo', 'bar']
+    result = pd.Index(index)
+    tm.assert_index_equal(
```
write the expected out on a separate line

jreback commented on pandas/tests/indexes/multi/test_conversion.py:
```diff
@@ -0,0 +1,176 @@
+    assert result.names == index.names
+
+
+@pytest.mark.skipif(PY3, reason="testing legacy pickles not support on py3")
```
there is an open issue to blow these legacy pickles away

jreback commented on pandas/tests/indexes/multi/test_conversion.py:
```diff
@@ -0,0 +1,176 @@
+    assert_almost_equal(exp, exp2)
+
+
+def test_legacy_v2_unpickle(datapath):
```
same

jreback commented on pandas/tests/indexes/multi/test_copy.py:
```diff
@@ -0,0 +1,124 @@
+    # gh-12309: Check that the "name" argument
+    # passed at initialization is honored.
+
+    # TODO: Remove or refactor MultiIndex not tested.
```
yep refactor this

elmq0022 replied:
I wrote a new test specifically for MultiIndexes. I test the levels, names, and the labels are honored when copied. I'm not sure about the dtype though. Looks like this is always an object anyway.

jreback commented on pandas/tests/indexes/multi/test_copy.py:
```diff
@@ -0,0 +1,124 @@
+    # GH12309
+    # TODO: REMOVE THIS TEST.  MultiIndex is tested seperately as noted below.
+
+    for name, index in compat.iteritems({'idx': idx}):
```
refactor this to avoid the iteration (and just use idx directly as a fixture)

elmq0022 replied:
I wrote a new test specifically for MultiIndexes.  I test the levels, names, and the labels are honored when copied.  I'm not sure about the dtype though.  Looks like this is always an object anyway.

elmq0022 replied:
wrote a new version below.

jreback commented on pandas/tests/indexes/multi/test_copy.py:
```diff
@@ -0,0 +1,124 @@
+
+def test_copy_and_deepcopy(indices):
+
+    if isinstance(indices, MultiIndex):
```
parameterize and don't need the fixure (it can just be a MI)

jreback commented on pandas/tests/indexes/multi/test_format.py:
```diff
@@ -0,0 +1,133 @@
+    tm.assert_index_equal(result, mi_u, exact=True)
+
+
+def test_str():
```
can remove / skip

jreback commented on pandas/tests/indexes/multi/test_get_set.py:
```diff
@@ -0,0 +1,423 @@
+    levels = idx.levels
+    new_levels = [[lev + 'a' for lev in level] for level in levels]
+
+    def assert_matching(actual, expected, check_dtype=False):
```
move to the top of the file (assert_matching)

jreback commented on pandas/tests/indexes/multi/test_get_set.py:
```diff
@@ -0,0 +1,423 @@
+    minor_labels = [(x + 1) % 1 for x in minor_labels]
+    new_labels = [major_labels, minor_labels]
+
+    def assert_matching(actual, expected):
```
e.g. its repeated

jreback commented on pandas/tests/indexes/multi/test_get_set.py:
```diff
@@ -0,0 +1,423 @@
+
+def test_set_levels_categorical():
+    # GH13854
+    index = MultiIndex.from_arrays([list("xyzx"), [0, 1, 2, 3]])
```
parmaeterize

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,448 @@
+    tm.assert_series_equal(left, right)
+
+
+def test_bounds(idx):
```
remove (and remove _bounds, I think we have an issue for this)

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,448 @@
+    tm.assert_dict_equal(groups, exp)
+
+
+def test_truncate():
```
move to test_analytics

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,448 @@
+    pytest.raises(ValueError, index.truncate, 3, 1)
+
+
+def test_where():
```
move *where* to test_analytics

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,448 @@
+                           idx.reorder_levels, [2, 1, 0])
+
+
+def test_astype(idx):
```
I think we have other astype tests somewhere else? if not, creata test_astype

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,448 @@
+            idx.astype('category')
+
+
+def test_repeat():
```
move to test_analytics

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,448 @@
+        first.tolist() - idx[-3:]
+
+
+def test_argsort(idx):
```
move to the sorting tests

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,448 @@
+    tm.assert_numpy_array_equal(result, expected)
+
+
+def test_map(idx):
```
move *map* tests to test_analytics

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,448 @@
+                    with np.errstate(all='ignore'):
+                        func(idx)
+
+    for func in [np.isfinite, np.isinf, np.isnan, np.signbit]:
```
want to split & parameterize these tests

jreback commented on pandas/tests/indexes/multi/test_operations.py:
```diff
@@ -0,0 +1,448 @@
+    tm.assert_index_equal(result, expected)
+
+
+def test_numpy_ufuncs(idx):
```
I would just combine test_operations and test_analytics (call it test_analytics)

jreback reviewed (approved):
bunch of comments. but let's do these in a follow up please. thanks!

jreback commented:
thanks @elmq0022 very very nice! great job!

if you can do a followup to address comments (and some TODO's you had in the code would be much appreciated)
